Mathlib/Order/Birkhoff.lean

```diff
@@ -0,0 +1,228 @@
+section DistribLattice
+variable [DistribLattice α]
+
+section Fintype
+variable [Fintype α] [OrderBot α]
+
+open scoped Classical
+
+/-- **Birkhoff's Representation Theorem**. Any nonempty finite distributive lattice is isomorphic to
+the lattice of lower sets of its sup-irreducible elements. -/
+noncomputable def OrderIso.lowerSetSupIrred : α ≃o LowerSet {a : α // SupIrred a} :=
+  Equiv.toOrderIso
+    { toFun := fun a ↦ ⟨{b | ↑b ≤ a}, fun b c hcb hba ↦ hba.trans' hcb⟩
+      invFun := fun s ↦ (s : Set {a : α // SupIrred a}).toFinset.sup (↑)
+      left_inv := fun a ↦ by
+        refine' le_antisymm (Finset.sup_le fun b ↦ Set.mem_toFinset.1) _
+        obtain ⟨s, rfl, hs⟩ := exists_supIrred_decomposition a
+        exact Finset.sup_le fun i hi ↦
+          le_sup_of_le (b := ⟨i, hs hi⟩) (Set.mem_toFinset.2 $ le_sup (f := id) hi) le_rfl
+      right_inv := fun s ↦ by
+        ext a
+        dsimp
+        refine' ⟨fun ha ↦ _, fun ha ↦ _⟩
+        · obtain ⟨i, hi, ha⟩ := a.2.supPrime.le_finset_sup.1 ha
+          exact s.lower ha (Set.mem_toFinset.1 hi)
+        · dsimp
+          exact le_sup (Set.mem_toFinset.2 ha) }
+    (fun b c hbc d ↦ le_trans' hbc) fun s t hst ↦ Finset.sup_mono $ Set.toFinset_mono hst
+
+attribute [-instance] Set.fintypeIic
+
+/-- Any nonempty finite distributive lattice is isomorphic to its lattice of sup-irreducible lower
+sets. -/
+noncomputable def OrderIso.supIrredLowerSet : α ≃o {s : LowerSet α // SupIrred s} :=
+  Equiv.toOrderIso
+    { toFun := fun a ↦ ⟨LowerSet.Iic a, LowerSet.supIrred_Iic _⟩
+      invFun := fun s ↦ ((s : LowerSet α) : Set α).toFinset.sup id
+      left_inv := fun a ↦ by
+        have : LocallyFiniteOrder α := Fintype.toLocallyFiniteOrder
+        simp
+      right_inv := by
+        classical
+        have : LocallyFiniteOrder α := Fintype.toLocallyFiniteOrder
+        rintro ⟨s, hs⟩
+        obtain ⟨a, rfl⟩ := LowerSet.supIrred.1 hs
+        simp }
+    (fun b c hbc d ↦ le_trans' hbc) fun s t hst ↦ Finset.sup_mono $ Set.toFinset_mono hst
+
+end Fintype
+
+variable (α)
+
+namespace OrderEmbedding
+variable [Fintype α] [@DecidablePred α SupIrred]
+
+/-- **Birkhoff's Representation Theorem**. Any finite distributive lattice can be embedded in a
+powerset lattice. -/
+noncomputable def birkhoffSet : α ↪o Set {a : α // SupIrred a} := by
+  by_cases IsEmpty α
+  · exact ⟨⟨isEmptyElim, isEmptyElim⟩, isEmptyElim ‹α›, isEmptyElim ‹α›⟩
+  rw [not_isEmpty_iff] at h
+  have := Fintype.toOrderBot α
+  exact OrderIso.lowerSetSupIrred.toOrderEmbedding.trans ⟨⟨_, SetLike.coe_injective⟩, Iff.rfl⟩
+
+/-- **Birkhoff's Representation Theorem**. Any finite distributive lattice can be embedded in a
+powerset lattice. -/
+noncomputable def birkhoffFinset : α ↪o Finset {a : α // SupIrred a} := by
+  exact (birkhoffSet _).trans Fintype.finsetOrderIsoSet.symm.toOrderEmbedding
+
+variable {α}
+
+@[simp] lemma coe_birkhoffFinset (a : α) : birkhoffFinset α a = birkhoffSet α a := by
+  classical
+  simp [birkhoffFinset]
```

Comment:
Could you make this a `simp only`? (Maybe with `says`?)

Reply:
This is supposed to be a single simp call, so I'm unwilling to squeeze it. I've left a todo instead.